### PR TITLE
CouchDB python module (from Apache CouchDB maintainers)

### PR DESCRIPTION
**Pull request by wohali**
This PR adds netdata support for Apache CouchDB 2.x.

Sample graphs: https://imgur.com/a/4Gz7q

The plugin is not strictly zero-config; users must specify a valid username/password pair, and they may have to set the node name via the node parameter as well. There's no easy way to auto-discover this information in CouchDB 2.x. (We're planning improvements for 3.x!)

I am on the Apache CouchDB PMC, a core developer, and I'd love to see this land in netdata soon! Please let me know if any changes are required.

### CODE REVIEW

**Comment by ktsaou** [comment]
Wow! This is nice. Thanks!

@l2isbad please review this plugin (no offense, we have to make sure this is optimal for netdata).

@wohali a few suggections:

1. you may want to add some information above the charts, possibly with links to your documentation. The idea is to provide key information, directly on the dashboard, to help the users diagnose issues or even find additional info while troubleshooting issues.

   This is done by editing this: https://github.com/firehol/netdata/blob/master/web/dashboard_info.js - There are 3 sections: the main menu (you can pick an icon here: http://fontawesome.io/icons/), the submenus and the charts.

   Generally, adding chart specific information to explain in plain English what the users are seeing at the chart, improves the effectiveness of netdata significantly. Many people read this info, especially when they face performance issues.

2. I see you have added one alarm that detects if netdata can collect metrics from couchdb. Are there any other conditions that users will have to be alerted? If you can describe any such cases, I could help creating alarms for them.

3. The plugin should also be [added to this page](https://github.com/firehol/netdata/wiki/Add-more-charts-to-netdata).

4. Also there should be some info to this [README.md](https://github.com/firehol/netdata/blob/master/python.d/README.md) and it should be linked at the wiki sidebar:

screenshot from 2017-09-30 00-27-18

---

I am adding here your logo, to be served by github CDN, for the [netdata infographic](http://my-netdata.io/infographic.html). 

image

If you want to edit the diagram yourself, it is [here](https://github.com/firehol/netdata/blob/master/diagrams/netdata-overview.xml), and it is [editable (load/save) directly by draw.io](https://www.draw.io/?title=netdata-overview.xml#Uhttps%3A%2F%2Fraw.githubusercontent.com%2Ffirehol%2Fnetdata%2Fmaster%2Fdiagrams%2Fnetdata-overview.xml) - edit the URL to replace `firehol` with your account and draw.io will open it from your github fork and add it to this PR when you save it.

If you edit the inforgraphic, please resize a few of the others to make room ( be generous to yourself :) and make sure you paste, at draw.io, the github image URL `https://user-images.githubusercontent.com/2662304/31036171-c62e6012-a573-11e7-92b1-d385911756fb.png`, not the image itself, so that the XML of the drawing remains small and the image comes from github.

If you don't want to bother with the infographic, I can add this later.

**Comment by wohali** [comment]
Thanks for the advice and nice words, @ktsaou ! @l2isbad looking forward to your code review.

I'll work on point 1 above over the next day or two, squash, force-push and comment here when the changes are ready.

For point 2, there's only one I can think of: `couchdb.httpd_status_codes.500.value` should always be `0`. Defining an alert whenever this occurs would be useful; the clear would be `x` minutes of no 500 errors, I guess. The problem is that I don't think that metric even exists until the value returns non-zero, am I correct? I know the graph doesn't get plotted until the value is non-zero at some time.

There's nothing further that we can think of to add for alerts at the present time. We're going to keep thinking about this one and will raise new PRs if we think of any. The big ones (e.g. out of disk space, CPU load, out of RAM) are already handled by netdata already.

For points 3 and 4, I'm happy to update the wiki myself, but shouldn't I wait until after this PR lands?

I'll also look at updating the main infographic and including that in the PR as well.

File: python.d/couchdb.chart.py [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+        self.definitions = CHARTS
+        self.host = self.configuration.get('host', '127.0.0.1')
+        self.port = self.configuration.get('port', 5984)
+        self.node = self.configuration.get('node', 'couchdb@127.0.0.1')
+        self.scheme = self.configuration.get('scheme', 'http')
+        self.user = self.configuration.get('user')
+        self.password = self.configuration.get('pass')
+        try:
+            self.dbs = self.configuration.get('databases').split(' ')
+        except Exception:
```
Comment (l2isbad) [inline python.d/couchdb.chart.py]: please be more specific on exception

File: python.d/couchdb.chart.py [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+            data = loads(raw_data)
+            for metric in stats:
+                value = data
+                metrics_list = metric.split('.')
+                try:
+                    for m in metrics_list:
+                        value = value[m]
+                except Exception as e:
+                    self.debug('cannot process ' + metrics_list + ' for ' + db
+                               + ": " + e)
```
Comment (l2isbad) [inline python.d/couchdb.chart.py]: convert `error` to `str` (better to use `str.format`) and too broad exception

File: python.d/couchdb.chart.py [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+                    continue
+                metric_name = 'db_{}_{}'.format(db, '_'.join(metrics_list))
```
Comment (l2isbad) [inline python.d/couchdb.chart.py]:
```python
Python 2.6.6
[GCC 4.4.5] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> '{}'.format('A')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: zero length field name in format
>>> '{0}'.format('A')
'A'
```

File: python.d/couchdb.chart.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+        except Exception as e:
+            self.debug('cannot process ' + metrics_list + ': ' + e)
+            continue
+        # strip off .value from end of stat
+        if metrics_list[-1] == 'value':
+            metrics_list = metrics_list[:-1]
+        # sum up 3xx/4xx/5xx
+        if metrics_list[0:2] == ['couchdb', 'httpd_status_codes'] and \
+                int(metrics_list[2]) > 202:
+            metrics_list[2] = '{}xx'.format(int(metrics_list[2]) // 100)
```
Comment (l2isbad) [inline python.d/couchdb.chart.py]: same here

File: python.d/couchdb.chart.py [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+
+def fetch_data(raw_data, metrics):
+    data = dict()
+    for metric in metrics:
+        value = raw_data
+        metrics_list = metric.split('.')
+        try:
+            for m in metrics_list:
+                value = value[m]
+        except Exception as e:
```
Comment (l2isbad) [inline python.d/couchdb.chart.py]: be more specific on exception

**Comment by wohali** [comment]
Update: I have updated the infographic, `python.d/README.md` and the `web/dashboard_info.js` file with suitable info.

I think we can wait on adding the 5xx alarm, it's not critical. I'd rather see this PR land sooner 😉 

Over to you!

File: python.d/couchdb.chart.py [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+        if 'message_queues' in data:
+            to_netdata['peak_msg_queue'] = get_peak_msg_queue(data)
+        return queue.put(to_netdata)
+
+    def _get_active_tasks_stats(self, queue, url, _):
+        taskdict = defaultdict(int)
+        taskdict["activetasks_indexer"] = 0
+        taskdict["activetasks_database_compaction"] = 0
+        taskdict["activetasks_replication"] = 0
+        taskdict["activetasks_view_compaction"] = 0
```
Comment (l2isbad) [inline python.d/couchdb.chart.py]: Why do we need to initialize these values? Can they not be in the `data`?

**Comment by l2isbad** [comment]
gj @wohali 😄

File: python.d/couchdb.chart.py (hunk 6 above)
Comment (wohali) [reply python.d/couchdb.chart.py]: Yes, sometimes they're not in the data; if there are no active tasks of that type, the entire stanza will be missing.

**Comment by wohali** [comment]
@l2isbad everything's been addressed, as far as I can tell. Let me know if it meets your standards!

**Comment by l2isbad** [comment]
Looks ok for me 👍 
Please post here few screenshots of current version + runtime chart (from Netdata Monitroing/python.d)
@ktsaou may want to add them to https://github.com/firehol/netdata/issues/2788

**Comment by l2isbad** [comment]
@shellshock1953 as far as I remember you are a couchdb user ^

**Comment by wohali** [comment]
As requested:
mwsnap 2017-09-29 22_54_33
mwsnap 2017-09-29 22_54_44
mwsnap 2017-09-29 22_54_56
mwsnap 2017-09-29 22_55_48
mwsnap 2017-09-29 22_55_54
mwsnap 2017-09-29 22_56_22
mwsnap 2017-09-29 23_02_40

**Comment by ktsaou** [comment]
This is really nice!

I am merging it. I need your help to write the alarm, so I will invite you on a new issue.

**Comment by ktsaou** [comment]
btw, remember to update the wiki

**Comment by wohali** [comment]
Done 👯  🍁 :canada: 

I need your help to update the main /README.md and the main wiki page with the new infographic, I'm not exactly sure how to get it hosted by github.

**Comment by ktsaou** [comment]
Oh... it is very simple: just export a PNG from draw.io (default settings), and upload it to a comment on this PR. Github will generate a URL for it. Then copy this URL and paste it at the other places.

**Comment by VDigitall** [comment]
@wohali, do you plan add support СouchDB 1.6.x?

**Comment by wohali** [comment]
@VDigitall No, nor 1.7.x, this is a 2.x plugin only.

(You should be upgrading ASAP, see https://blog.couchdb.org/2017/11/14/apache-couchdb-cve-2017-12635-and-cve-2017-12636/ , at least to 1.7.1 if not 2.x.)

**Comment by l2isbad** [comment]
@wohali 
https://github.com/firehol/netdata/blob/master/python.d/couchdb.chart.py#L181
i think we should convert values to KB

**Pull request merged**